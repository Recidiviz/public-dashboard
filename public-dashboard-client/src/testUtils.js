// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import fs from "fs";
import path from "path";
import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components/macro";
import { THEME } from "./theme";
import { InfoPanelProvider } from "./info-panel";

// eslint-disable-next-line react/prop-types
const GlobalWrapper = ({ children }) => {
  // include globally expected Context providers or other required wrappers here
  return (
    <ThemeProvider theme={THEME}>
      <InfoPanelProvider>{children}</InfoPanelProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalWrapper, ...options });

// re-export everything from testing-library
export * from "@testing-library/react";

// override render method
export { customRender as render };

// provide original render method as a fallback if needed
export { render as renderUnwrapped };

// retrieve a data fixture from spotlight-api
export function getDataFixture(filename) {
  const filePath = path.resolve(
    __dirname,
    `../../spotlight-api/core/demo_data/${filename}`
  );
  const stringContents = fs.readFileSync(filePath).toString();

  // copypasta from spotlight-api/core/metricsApi for transforming JSONLines
  if (!stringContents || stringContents.length === 0) {
    return null;
  }
  const jsonObject = [];
  const splitStrings = stringContents.split("\n");
  splitStrings.forEach((line) => {
    if (line) {
      jsonObject.push(JSON.parse(line));
    }
  });

  return jsonObject;
}
