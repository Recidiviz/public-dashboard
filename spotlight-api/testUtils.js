// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

const path = require("path");
const ReadLines = require("n-readlines");

/**
 * Parses and returns the first line of the JSONLines fixture,
 * for use as a proxy for correct filesystem data fetching
 */
function getFirstRecordFromFixture(metricName) {
  const lineReader = new ReadLines(
    path.resolve(__dirname, `./core/demo_data/${metricName}`)
  );
  return JSON.parse(lineReader.next().toString());
}

const expectedMetricsByGroup = {
  parole: [
    "active_program_participation_by_region",
    "site_offices",
    "supervision_population_by_district_by_demographics",
    "supervision_revocations_by_period_by_type_by_demographics",
    "supervision_success_by_month",
    "supervision_success_by_period_by_demographics",
    "supervision_population_by_month_by_demographics",
  ],
  prison: [
    "incarceration_facilities",
    "incarceration_population_by_admission_reason",
    "incarceration_population_by_facility_by_demographics",
    "incarceration_releases_by_type_by_period",
    "incarceration_lengths_by_demographics",
    "incarceration_population_by_month_by_demographics",
    "recidivism_rates_by_cohort_by_year",
  ],
  probation: [
    "active_program_participation_by_region",
    "judicial_districts",
    "supervision_population_by_district_by_demographics",
    "supervision_revocations_by_period_by_type_by_demographics",
    "supervision_success_by_month",
    "supervision_success_by_period_by_demographics",
    "supervision_population_by_month_by_demographics",
  ],
  race: ["racial_disparities"],
  sentencing: [
    "judicial_districts",
    "sentence_type_by_district_by_demographics",
  ],
};

module.exports = {
  getFirstRecordFromFixture,
  expectedMetricsByGroup,
};
