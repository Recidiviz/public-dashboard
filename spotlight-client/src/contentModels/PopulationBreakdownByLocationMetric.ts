// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { sum } from "d3-array";
import { computed, makeObservable } from "mobx";
import {
  DemographicView,
  DemographicViewList,
  getDemographicCategories,
  getDemographicViewLabel,
  recordIsTotalByDimension,
} from "../demographics";
import {
  PopulationBreakdownByLocationRecord,
  recordMatchesLocality,
} from "../metricsApi";
import { colors } from "../UiLibrary";
import calculatePct from "./calculatePct";
import { countUnknowns, hasUnknowns } from "./unknowns";
import Metric, { BaseMetricConstructorOptions } from "./Metric";
import { DemographicCategoryRecords, UnknownCounts } from "./types";

export default class PopulationBreakdownByLocationMetric extends Metric<
  PopulationBreakdownByLocationRecord
> {
  readonly totalLabel: string;

  constructor(
    props: BaseMetricConstructorOptions<PopulationBreakdownByLocationRecord> & {
      totalLabel: string;
    }
  ) {
    super(props);

    this.totalLabel = props.totalLabel;

    makeObservable(this, {
      dataSeries: computed,
      totalPopulation: computed,
      unknowns: computed,
    });
  }

  get records(): PopulationBreakdownByLocationRecord[] | undefined {
    let recordsToReturn = this.allRecords;
    if (!recordsToReturn) return undefined;

    recordsToReturn = recordsToReturn.filter(
      recordMatchesLocality(this.localityId)
    );

    return recordsToReturn;
  }

  get dataSeries(): DemographicCategoryRecords[] | null {
    const { records } = this;
    if (!records) return null;

    return DemographicViewList.filter(
      (view): view is Exclude<DemographicView, "total" | "nofilter"> =>
        view !== "total" && view !== "nofilter"
    ).map((demographicView) => {
      return {
        label: getDemographicViewLabel(demographicView),
        records: calculatePct(
          getDemographicCategories(demographicView).map(
            ({ identifier, label }, index) => {
              let value = 0;
              const matchingRecord = records.find(
                (record) => record[demographicView] === identifier
              );
              if (matchingRecord) {
                value = matchingRecord.population;
              }

              return {
                color: colors.dataViz[index],
                label,
                value,
              };
            }
          )
        ),
      };
    });
  }

  get totalPopulation(): number | undefined {
    const { records } = this;
    if (!records) return undefined;

    const totalRecord = records.find(recordIsTotalByDimension("total"));
    if (!totalRecord) return undefined;

    return totalRecord.population;
  }

  get unknowns(): UnknownCounts | undefined {
    const { records } = this;

    if (!records) return undefined;

    const counts = countUnknowns(
      records,
      (groupedRecords: PopulationBreakdownByLocationRecord[]) =>
        sum(groupedRecords, (r) => r.population)
    );

    return hasUnknowns(counts) ? counts : undefined;
  }
}
