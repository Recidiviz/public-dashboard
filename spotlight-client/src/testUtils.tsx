// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { LocationProvider } from "@reach/router";
import { render } from "@testing-library/react";
import React from "react";
import { autorun } from "mobx";
import waitForLocalhost from "wait-for-localhost";

// eslint-disable-next-line import/prefer-default-export
export function waitForTestServer(): Promise<void> {
  return waitForLocalhost({ path: "/health", port: 3002 });
}

/**
 * Convenience method to run an immediate, one-time reactive effect
 */
export function reactImmediately(effect: () => void): void {
  // this will call the effect function immediately,
  // and then immediately call the disposer to tear down the reaction
  autorun(effect)();
}

const LocationContextWrapper: React.FC = ({ children }) => {
  return <LocationProvider>{children}</LocationProvider>;
};

/**
 * Convenience method for rendering components that use @reach/router hooks
 * in a LocationContext to prevent rendering errors.
 */
export const renderWithRouter = (
  ui: React.ReactElement
): ReturnType<typeof render> => {
  return render(ui, { wrapper: LocationContextWrapper });
};
