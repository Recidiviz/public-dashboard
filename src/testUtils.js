import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { THEME } from "./constants";

// eslint-disable-next-line react/prop-types
const GlobalWrapper = ({ children }) => {
  // include globally expected Context providers or other required wrappers here
  return <ThemeProvider theme={THEME}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalWrapper, ...options });

// re-export everything from testing-library
export * from "@testing-library/react";

// override render method
export { customRender as render };

// provide original render method as a fallback if needed
export { render as renderUnwrapped };
