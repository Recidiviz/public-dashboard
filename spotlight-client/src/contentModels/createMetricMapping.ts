// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { assertNever } from "assert-never";
import { MetricTypeIdList, TenantContent, TenantId } from "../contentApi/types";
import {
  parolePopulationCurrent,
  parolePopulationHistorical,
  paroleProgramParticipationCurrent,
  paroleRevocationReasons,
  paroleSuccessRateDemographics,
  paroleSuccessRateMonthly,
  prisonAdmissionReasons,
  prisonPopulationCurrent,
  prisonPopulationHistorical,
  prisonReleaseTypes,
  prisonStayLengths,
  probationPopulationCurrent,
  probationPopulationHistorical,
  probationProgramParticipationCurrent,
  probationRevocationReasons,
  probationSuccessRateDemographics,
  probationSuccessRateMonthly,
  recidivismRateAllFollowup,
  recidivismRateConventionalFollowup,
  sentencePopulationCurrent,
  sentenceTypesCurrent,
} from "../metricsApi";
import { MetricMapping } from "./types";
import DemographicsByCategoryMetric from "./DemographicsByCategoryMetric";
import HistoricalPopulationBreakdownMetric from "./HistoricalPopulationBreakdownMetric";
import PopulationBreakdownByLocationMetric from "./PopulationBreakdownByLocationMetric";
import ProgramParticipationCurrentMetric from "./ProgramParticipationCurrentMetric";
import RecidivismRateMetric from "./RecidivismRateMetric";
import SentenceTypeByLocationMetric from "./SentenceTypeByLocationMetric";
import SupervisionSuccessRateDemographicsMetric from "./SupervisionSuccessRateDemographicsMetric";
import SupervisionSuccessRateMonthlyMetric from "./SupervisionSuccessRateMonthlyMetric";
import { ERROR_MESSAGES } from "../constants";
import { NOFILTER_KEY, TOTAL_KEY } from "../demographics";
import { colors } from "../UiLibrary";

type MetricMappingFactoryOptions = {
  localityLabelMapping: TenantContent["localities"];
  metadataMapping: TenantContent["metrics"];
  tenantId: TenantId;
};
/**
 * Factory function for converting a mapping of content objects by metric ID
 * to a mapping of Metric instances by metric ID. Creating the entire mapping at once
 * ensures that each ID maps to the proper Metric type without requiring further
 * type guarding on the part of consumers.
 */
export default function createMetricMapping({
  localityLabelMapping,
  metadataMapping,
  tenantId,
}: MetricMappingFactoryOptions): MetricMapping {
  const metricMapping: MetricMapping = new Map();

  // to maintain type safety we iterate through all of the known metrics;
  // iterating through the metadata object's keys widens the type to `string`,
  // which prevents us from guaranteeing exhaustiveness at the type level
  MetricTypeIdList.forEach((metricType) => {
    // not all metrics are required; metadata object is the source of truth
    // for which metrics to include
    const metadata = metadataMapping[metricType];
    if (!metadata) {
      return;
    }

    // messages for errors we may hit at various points below
    const localityContentError = `${ERROR_MESSAGES.missingRequiredContent} (locality labels)`;
    const totalLabelError = `${ERROR_MESSAGES.missingRequiredContent} (total population label)`;

    // this big ol' switch statement ensures that the type ID string union is properly narrowed,
    // allowing for 1:1 correspondence between the ID and the typed Metric instance.
    switch (metricType) {
      case "SentencePopulationCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Sentencing,
              dataTransformer: sentencePopulationCurrent,
              sourceFileName: "sentence_type_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);

        break;
      case "SentenceTypesCurrent":
        if (!localityLabelMapping?.Sentencing)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SentenceTypeByLocationMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Sentencing,
            dataTransformer: sentenceTypesCurrent,
            sourceFileName: "sentence_type_by_district_by_demographics",
          })
        );
        break;
      case "PrisonPopulationCurrent":
        if (!localityLabelMapping?.Prison)
          throw new Error(localityContentError);

        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Prison,
              dataTransformer: prisonPopulationCurrent,
              sourceFileName:
                "incarceration_population_by_facility_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ProbationPopulationCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Probation,
              dataTransformer: probationPopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "ParolePopulationCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);
        if ("totalLabel" in metadata)
          metricMapping.set(
            metricType,
            new PopulationBreakdownByLocationMetric({
              ...metadata,
              id: metricType,
              tenantId,
              defaultDemographicView: NOFILTER_KEY,
              defaultLocalityId: TOTAL_KEY,
              localityLabels: localityLabelMapping.Parole,
              dataTransformer: parolePopulationCurrent,
              sourceFileName:
                "supervision_population_by_district_by_demographics",
            })
          );
        else throw new Error(totalLabelError);
        break;
      case "PrisonPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonPopulationHistorical,
            sourceFileName: "incarceration_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationPopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationPopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ParolePopulationHistorical":
        metricMapping.set(
          metricType,
          new HistoricalPopulationBreakdownMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: parolePopulationHistorical,
            sourceFileName: "supervision_population_by_month_by_demographics",
          })
        );
        break;
      case "ProbationProgrammingCurrent":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ParoleProgrammingCurrent":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new ProgramParticipationCurrentMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: NOFILTER_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleProgramParticipationCurrent,
            sourceFileName: "active_program_participation_by_region",
          })
        );
        break;
      case "ProbationSuccessHistorical":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ParoleSuccessHistorical":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateMonthlyMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: undefined,
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateMonthly,
            sourceFileName: "supervision_success_by_month",
          })
        );
        break;
      case "ProbationSuccessAggregate":
        if (!localityLabelMapping?.Probation)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Probation,
            dataTransformer: probationSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ParoleSuccessAggregate":
        if (!localityLabelMapping?.Parole)
          throw new Error(localityContentError);

        metricMapping.set(
          metricType,
          new SupervisionSuccessRateDemographicsMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: TOTAL_KEY,
            localityLabels: localityLabelMapping.Parole,
            dataTransformer: paroleSuccessRateDemographics,
            sourceFileName: "supervision_success_by_period_by_demographics",
          })
        );
        break;
      case "ProbationRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: probationRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "ParoleRevocationsAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: paroleRevocationReasons,
            sourceFileName:
              "supervision_revocations_by_period_by_type_by_demographics",
          })
        );
        break;
      case "PrisonAdmissionReasonsCurrent":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonAdmissionReasons,
            sourceFileName: "incarceration_population_by_admission_reason",
          })
        );
        break;
      case "PrisonReleaseTypeAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonReleaseTypes,
            sourceFileName: "incarceration_releases_by_type_by_period",
          })
        );
        break;
      case "PrisonRecidivismRateHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateAllFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonRecidivismRateSingleFollowupHistorical":
        metricMapping.set(
          metricType,
          new RecidivismRateMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            followUpYears: 3,
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: recidivismRateConventionalFollowup,
            sourceFileName: "recidivism_rates_by_cohort_by_year",
          })
        );
        break;
      case "PrisonStayLengthAggregate":
        metricMapping.set(
          metricType,
          new DemographicsByCategoryMetric({
            ...metadata,
            id: metricType,
            tenantId,
            defaultDemographicView: "total",
            defaultLocalityId: undefined,
            localityLabels: undefined,
            dataTransformer: prisonStayLengths,
            sourceFileName: "incarceration_lengths_by_demographics",
            color: colors.dataVizNamed.get("teal"),
          })
        );
        break;
      default:
        assertNever(metricType);
    }
  });

  return metricMapping;
}
