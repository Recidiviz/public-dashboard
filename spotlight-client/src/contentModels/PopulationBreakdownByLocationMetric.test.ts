// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import fetchMock from "jest-fetch-mock";
import { runInAction, when } from "mobx";
import { reactImmediately } from "../testUtils";
import createMetricMapping from "./createMetricMapping";
import PopulationBreakdownByLocationMetric from "./PopulationBreakdownByLocationMetric";
import contentFixture from "./__fixtures__/tenant_content_exhaustive";

const testTenantId = "US_ND";
const testMetricId = "PrisonPopulationCurrent";
const testMetadataMapping = {
  [testMetricId]: contentFixture.metrics[testMetricId],
};

function getTestMetric() {
  return createMetricMapping({
    localityLabelMapping: contentFixture.localities,
    metadataMapping: testMetadataMapping,
    tenantId: testTenantId,
    demographicFilter: contentFixture.demographicCategories,
  }).get(testMetricId) as PopulationBreakdownByLocationMetric;
}

test("locality filter", async () => {
  const metric = getTestMetric();

  metric.hydrate();

  await when(() => metric.records !== undefined);

  reactImmediately(() =>
    expect(metric.records?.every((record) => record.locality === "ALL")).toBe(
      true
    )
  );

  const facilityId = contentFixture.localities.Prison.entries[1].id;

  runInAction(() => {
    metric.localityId = facilityId;
  });

  reactImmediately(() => {
    expect(metric.records?.length).toBeGreaterThan(0);
    expect(
      metric.records?.every((record) => record.locality === facilityId)
    ).toBe(true);
  });

  expect.hasAssertions();
});

test("demographic data series", async () => {
  const metric = getTestMetric();

  metric.hydrate();

  await when(() => Boolean(metric.dataSeries));

  reactImmediately(() => {
    expect(metric.dataSeries).toMatchSnapshot();
  });

  expect.hasAssertions();
});

test("total population", async () => {
  const metric = getTestMetric();

  metric.hydrate();

  await when(() => Boolean(metric.records));

  reactImmediately(() => expect(metric.totalPopulation).toBe(2041));

  const facilityId = contentFixture.localities.Prison.entries[1].id;

  runInAction(() => {
    metric.localityId = facilityId;
  });

  reactImmediately(() => expect(metric.totalPopulation).toBe(413));

  expect.hasAssertions();
});

test("no unknowns", async () => {
  const metric = getTestMetric();

  await metric.hydrate();

  reactImmediately(() => {
    expect(metric.unknowns).toBeUndefined();
  });

  expect.hasAssertions();
});

test("report unknowns for current locality", (done) => {
  const metric = getTestMetric();

  // mock unknowns in response
  fetchMock.mockOnce(
    JSON.stringify({
      incarceration_population_by_facility_by_demographics: [
        {
          state_code: "US_ND",
          date_of_stay: "2020-07-23",
          facility: "ALL",
          race_or_ethnicity: "EXTERNAL_UNKNOWN",
          gender: "ALL",
          age_bucket: "ALL",
          total_population: "25",
        },
        {
          state_code: "US_ND",
          date_of_stay: "2020-07-23",
          facility: "test1",
          race_or_ethnicity: "EXTERNAL_UNKNOWN",
          gender: "ALL",
          age_bucket: "ALL",
          total_population: "20",
        },
        {
          state_code: "US_ND",
          date_of_stay: "2020-07-23",
          facility: "test2",
          race_or_ethnicity: "EXTERNAL_UNKNOWN",
          gender: "ALL",
          age_bucket: "ALL",
          total_population: "5",
        },
      ],
    })
  );

  metric.hydrate();

  when(
    () => metric.unknowns !== undefined,
    () => {
      expect(metric.unknowns).toEqual({
        raceOrEthnicity: 25,
        gender: 0,
        ageBucket: 0,
      });

      runInAction(() => {
        metric.localityId = "test2";
      });

      expect(metric.unknowns).toEqual({
        raceOrEthnicity: 5,
        gender: 0,
        ageBucket: 0,
      });
      done();
    }
  );
});
