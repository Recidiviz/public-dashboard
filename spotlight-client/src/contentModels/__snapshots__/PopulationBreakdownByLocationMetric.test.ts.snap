// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demographic data series 1`] = `
Array [
  Object {
    "label": "Race or Ethnicity",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Native American",
        "pct": 0.21291448516579406,
        "value": 244,
      },
      Object {
        "color": "#D9A95F",
        "label": "Black",
        "pct": 0.08376963350785341,
        "value": 96,
      },
      Object {
        "color": "#BA4F4F",
        "label": "Hispanic",
        "pct": 0.10732984293193717,
        "value": 123,
      },
      Object {
        "color": "#4C6290",
        "label": "White",
        "pct": 0.5855148342059336,
        "value": 671,
      },
      Object {
        "color": "#90AEB5",
        "label": "Other",
        "pct": 0.010471204188481676,
        "value": 12,
      },
    ],
  },
  Object {
    "label": "Gender",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "Male",
        "pct": 0.907967032967033,
        "value": 661,
      },
      Object {
        "color": "#D9A95F",
        "label": "Female",
        "pct": 0.09203296703296704,
        "value": 67,
      },
    ],
  },
  Object {
    "label": "Age Group",
    "records": Array [
      Object {
        "color": "#25636F",
        "label": "<25",
        "pct": 0.13619402985074627,
        "value": 146,
      },
      Object {
        "color": "#D9A95F",
        "label": "25-29",
        "pct": 0.26492537313432835,
        "value": 284,
      },
      Object {
        "color": "#BA4F4F",
        "label": "30-39",
        "pct": 0.35074626865671643,
        "value": 376,
      },
      Object {
        "color": "#4C6290",
        "label": "40-49",
        "pct": 0.04197761194029851,
        "value": 45,
      },
      Object {
        "color": "#90AEB5",
        "label": "50-59",
        "pct": 0.061567164179104475,
        "value": 66,
      },
      Object {
        "color": "#CC989C",
        "label": "60-69",
        "pct": 0.09701492537313433,
        "value": 104,
      },
      Object {
        "color": "#B6CC98",
        "label": "70+",
        "pct": 0.04757462686567164,
        "value": 51,
      },
    ],
  },
]
`;
